From 27532950d4009664c3885591a6c6bdd42b367c31 Mon Sep 17 00:00:00 2001
From: Nicolas Fella <nicolas.fella@gmx.de>
Date: Mon, 14 Jun 2021 20:28:24 +0200
Subject: [PATCH] Bundle plugins on Android

On Android we need special handling for loading plugins. Usually that is done behind the scenes in KPluginLoader, but we don't use that here
---
 src/CMakeLists.txt                        |  7 +++++
 src/KF5Kirigami2-android-dependencies.xml |  9 +++++++
 src/libkirigami/platformtheme.cpp         | 32 ++++++++++++++++-------
 3 files changed, 38 insertions(+), 10 deletions(-)
 create mode 100644 src/KF5Kirigami2-android-dependencies.xml

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index bb14e287..01be720a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -136,3 +136,10 @@ else()
 endif()
 
 install(TARGETS kirigamiplugin DESTINATION ${KDE_INSTALL_QMLDIR}/org/kde/kirigami.2)
+
+if (ANDROID)
+    install(FILES KF5Kirigami2-android-dependencies.xml
+        DESTINATION ${KDE_INSTALL_LIBDIR}
+        RENAME KF5Kirigami2_${CMAKE_ANDROID_ARCH_ABI}-android-dependencies.xml
+    )
+endif()
diff --git a/src/KF5Kirigami2-android-dependencies.xml b/src/KF5Kirigami2-android-dependencies.xml
new file mode 100644
index 00000000..c494299d
--- /dev/null
+++ b/src/KF5Kirigami2-android-dependencies.xml
@@ -0,0 +1,9 @@
+<rules>
+    <dependencies>
+        <lib name="KF5Kirigami2">
+            <depends>
+                <bundled file="plugins/kf5/kirigami" />
+            </depends>
+        </lib>
+    </dependencies>
+</rules>
diff --git a/src/libkirigami/platformtheme.cpp b/src/libkirigami/platformtheme.cpp
index 3a44e769..6747f701 100644
--- a/src/libkirigami/platformtheme.cpp
+++ b/src/libkirigami/platformtheme.cpp
@@ -857,21 +857,33 @@ PlatformTheme *PlatformTheme::qmlAttachedProperties(QObject *object)
 #else
         const auto libraryPaths = QCoreApplication::libraryPaths();
         for (const QString &path : libraryPaths) {
+#ifdef Q_OS_ANDROID
+            QDir dir(path);
+#else
             QDir dir(path + QStringLiteral("/kf5/kirigami"));
+#endif
             const auto fileNames = dir.entryList(QDir::Files);
+
             for (const QString &fileName : fileNames) {
-                // TODO: env variable?
-                if (!QQuickStyle::name().isEmpty() && fileName.startsWith(QQuickStyle::name())) {
-                    QPluginLoader loader(dir.absoluteFilePath(fileName));
-                    QObject *plugin = loader.instance();
-                    // TODO: load actually a factory as plugin
-
-                    KirigamiPluginFactory *factory = qobject_cast<KirigamiPluginFactory *>(plugin);
-                    if (factory) {
-                        PlatformThemePrivate::s_pluginFactory = factory;
-                        return factory->createPlatformTheme(object);
+
+#ifdef Q_OS_ANDROID
+                if (fileName.startsWith(QStringLiteral("libplugins_kf5_kirigami_")) && QLibrary::isLibrary(fileName)) {
+#endif
+                    // TODO: env variable?
+                    if (!QQuickStyle::name().isEmpty() && fileName.contains(QQuickStyle::name())) {
+                        QPluginLoader loader(dir.absoluteFilePath(fileName));
+                        QObject *plugin = loader.instance();
+                        // TODO: load actually a factory as plugin
+
+                        KirigamiPluginFactory *factory = qobject_cast<KirigamiPluginFactory *>(plugin);
+                        if (factory) {
+                            PlatformThemePrivate::s_pluginFactory = factory;
+                            return factory->createPlatformTheme(object);
+                        }
                     }
+#ifdef Q_OS_ANDROID
                 }
+#endif
             }
         }
 #endif
-- 
2.32.0

